modules/grader/base.go

```diff
@@ -2,8 +2,31 @@ package grader
 
 import (
 	"fmt"
+	"os"
+	"strconv"
+	"sync"
+
+	lxr "github.com/pegnet/LXRHash"
 )
 
+// LX holds an instance of lxrhash
+var LX *lxr.LXRHash
+var lxInitializer sync.Once
+
+// The init function for LX is expensive. So we should explicitly call the init if we intend
+// to use it. Make the init call idempotent
+func InitLX() {
+	lxInitializer.Do(func() {
+		// This code will only be executed ONCE, no matter how often you call it
+		if size, err := strconv.Atoi(os.Getenv("LXRBITSIZE")); err == nil && size >= 8 && size <= 30 {
+			LX = lxr.Init(lxr.Seed, uint64(size), lxr.HashSize, lxr.Passes)
+		} else {
+			LX = lxr.Init(lxr.Seed, lxr.MapSizeBits, lxr.HashSize, lxr.Passes)
+		}
+
+	})
+}
```

Comment:
This is how we did it before. It's not a bad way to handle it, but still feels a bit off

Reply:
with the singletons, we don't need the `once` functionality anymore. just:
```go
var LX *lxr.LXRHash
func InitLX() {
	if size, err := strconv.Atoi(os.Getenv("LXRBITSIZE")); err == nil && size >= 8 && size <= 30 {
		LX = lxr.Init(lxr.Seed, uint64(size), lxr.HashSize, lxr.Passes)
	} else {
		LX = lxr.Init(lxr.Seed, lxr.MapSizeBits, lxr.HashSize, lxr.Passes)
	}
}

Reply:
True, but if we do without the `once`, then your repo that tracks the number of references will be a bit incorrect.

If we drop the `once`, then we should make a habit of always releasing it after use. Which I'm not sure is really worth it. I would feel a bit off creating multiple references to different LXR hashes.

With the sync wrapper, it's also threadsafe.

Reply:
pegnet wants lxrhash loaded in memory for the entire lifetime and we have no plans on ever releasing it. we're not using the module's release feature so we can create as many references as we want

Reply:
Yes. Maybe we should change it so you can pull in the modules repo, and release it yourself. 

Like as an importer, you can `modules.Release()` or w/e. Idk if we should force the caller to call the init? But I feel weird just creating a new reference each time if we have the release feature in the lxr repo. We can't guarantee the caller will want it for it's entire lifetime for any other application.

Reply:
i'm not quite sure what you meant

the singletons are only inside a single application, ie how pegnet manages the singletons has no effect on any other application.

the release functionality only exists for applications that want to temporarily load lxrhash (like a wallet that only needs it once an hour or something). they're going to have to design their application around that and make sure to release their references

pegnet needs the lxr hash until the app shuts down, so we never want to release it in the pegnet app, hence we can just keep calling lxr.Init(...) over and over if we want to. every call returns the same reference

Reply:
update: i think i get it now. we need a way for **other applications** who are using **this module** to release **this module's instance** of lxrhash

we can either: add an explicit Release() function OR auto-release. let's discuss it in the other PR